lib/db/schema/patch-029-30.sql

```diff
@@ -0,0 +1,32 @@
+-- Update deleteSessionToken stored procedure to delete from devices.
+CREATE PROCEDURE `deleteSessionToken_3` (
+  IN `tokenIdArg` BINARY(32)
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  DELETE FROM sessionTokens WHERE tokenId = tokenIdArg;
+  DELETE FROM unverifiedTokens WHERE tokenId = tokenIdArg;
+  DELETE FROM devices WHERE sessionTokenId = tokenIdArg;
+
+  COMMIT;
+END;
+
+-- Prior to this migration, it was possible to delete a session token
+-- without deleting the associated device (if a device record exists).
+-- That's fixed by deleteSessionTokens_3, but we should also purge any
+-- historical devices that have reached this state.
+DELETE FROM devices
+  USING devices LEFT JOIN sessionTokens
+    ON devices.sessionTokenId = sessionTokens.tokenId
+  WHERE sessionTokens.tokenId IS NULL;
```

Comment:
Yep, realise it needs @jrgm not to freak out in order to stay in, wanted to see what you thought of the concept first.

I'll try it out on my session-tokeny local db today, although given your point about cliffs and MySQL yesterday, it may not be proving much. I only have ~8 million session tokens at the moment, obviously I can keep increasing that but I've no idea at what point the cliff begins.